File: SortingAlgorithmsCpp/SortingAlgorithms/HeapSort.cpp
#include "HeapSort.hpp"

void swap(int* a, int* b) {
	int temp = *a;
	*a = *b;
	*b = temp;
}

void heapify(int arr[], int size, int i) {
	int largest = i;
	int l = 2 * i + 1;
	int r = 2 * i + 2;

	if (l < size && arr[l] > arr[largest]) {
		largest = l;
	}
	if (r < size && arr[r] > arr[largest]) {
		largest = r;
	}

	if (largest != i) {
		swap(&arr[i], &arr[largest]);
		heapify(arr, size, largest);
	}
}

void heapSort(int arr[], int size) {
	for (int i = size / 2 - 1; i >= 0; i--) {
		heapify(arr, size, i);
	}

	for (int i = size - 1; i >= 0; i--) {
		swap(&arr[0], &arr[i]);
		
		heapify(arr, i, 0);
	}
}

File: SortingAlgorithmsPython/heapSort.py
def heapify(arr, size, i):
    largest = i
    l = 2 * i + 1
    r = 2 * i + 2

    if l < size and arr[l] > arr[largest]:
        largest = l

    if r < size and arr[r] > arr[largest]:
        largest = r

    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, size, largest)

def heap_sort(arr):
    for i in range(len(arr) // 2 - 1, -1, -1):
        heapify(arr, len(arr), i)

    for i in range(len(arr) - 1, -1, -1):
        arr[0], arr[i] = arr[i], arr[0]
        heapify(arr, i, 0)